File: mobile/src/main/java/org/ligi/vaporizercontrol/ui/DataDisplayActivity.java
package org.ligi.vaporizercontrol.ui;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;
import android.widget.Toast;

import net.steamcrafted.loadtoast.LoadToast;

import org.ligi.tracedroid.sending.TraceDroidEmailSender;
import org.ligi.vaporizercontrol.R;
import org.ligi.vaporizercontrol.VaporizerDataBinder;
import org.ligi.vaporizercontrol.model.VaporizerData;
import org.ligi.vaporizercontrol.wiring.App;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnLongClick;
import fr.nicolaspomepuy.discreetapprate.AppRate;
import fr.nicolaspomepuy.discreetapprate.RetryPolicy;
import snow.skittles.Skittle;
import snow.skittles.SkittleBuilder;
import snow.skittles.SkittleContainer;
import snow.skittles.SkittleLayout;
import snow.skittles.TextSkittle;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

public class DataDisplayActivity extends AppCompatActivity implements VaporizerData.VaporizerUpdateListener {

    @Bind(R.id.intro_text)
    TextView introText;

    @Bind(R.id.skittleLayout)
    SkittleLayout fam;

    private LoadToast loadToast;

    private VaporizerDataBinder vaporizerDataBinder;

    @OnClick(R.id.led)
    void ledClick() {
        final boolean isUnknownOrNotBright = getApp().getVaporizerCommunicator().getData().ledPercentage == null ||
                getApp().getVaporizerCommunicator().getData().ledPercentage == 0;
        getApp().getVaporizerCommunicator().setLEDBrightness(isUnknownOrNotBright ? 100 : 0);
    }


    @OnLongClick(R.id.led)
    boolean onLEDLongClick() {
        ChangeDialogs.showLEDPercentageDialog(this, getApp().getVaporizerCommunicator());
        return true;
    }

    @OnClick({R.id.temperature, R.id.temperatureSetPoint, R.id.tempBoost})
    void onTemperatureClick() {
        ChangeDialogs.showTemperatureDialog(this, getApp().getVaporizerCommunicator());
    }

    //foo
    void onFAMClick() {
        //foofam.toggle();
    }

    //foo
    void editBoostClick() {
        ChangeDialogs.setBooster(this, getApp().getVaporizerCommunicator());
        //foofam.collapse();
    }

    //foo
    void editSetTempClick() {
        ChangeDialogs.showTemperatureDialog(this, getApp().getVaporizerCommunicator());
        //fam.collapse();
    }


    //foo
    void editLEDClick() {
        ChangeDialogs.showLEDPercentageDialog(this, getApp().getVaporizerCommunicator());
        //foofam.collapse();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        final SkittleBuilder builder = SkittleBuilder.newInstance(fam);
        builder.changeMainSkittleColor(getResources().getColor(R.color.accent));

        final TextSkittle.Builder foo = new TextSkittle.Builder("foo", getResources().getColor(R.color.accent), getResources().getDrawable(R.drawable.ic_action_bulb));

        foo.setTextBackground(getResources().getColor(R.color.accent));

        builder.addSkittle(foo.build());

//        builder.makeTextSkittle("foo",R.drawable.ic_action_bulb).add();

//        builder.addSkittle(R.drawable.ic_action_bulb, R.color.accent);

/*
        builder.setSkittleListener(new SkittleBuilder.SkittleClickListener() {
            @Override
            public void onSkittleClick(Skittle skittle) {

            }

            @Override
            public void onTextSkittleClick(TextSkittle textSkittle, String type) {

            }
        });
        builder.addSkittle(R.drawable.barratheon_icon, R.color.barratheon);
        builder.addSkittle(R.drawable.stark_icon, R.color.stark);
        */

        vaporizerDataBinder = new VaporizerDataBinder(this, getApp().getSettings());
        AppRate.with(this).retryPolicy(RetryPolicy.EXPONENTIAL).initialLaunchCount(5).checkAndShow();

        loadToast = new LoadToast(this);

        if (!TraceDroidEmailSender.sendStackTraces("ligi@ligi.de", this) && !getApp().getVaporizerCommunicator().getData().hasData()) {
            loadToast.setText("searching crafty");
            loadToast.show();
        }
    }


    @Override
    protected void onResume() {
        super.onResume();
        if (!getApp().getVaporizerCommunicator().isBluetoothAvailable()) {
            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    Toast.makeText(DataDisplayActivity.this, "can not scan - no BT available", Toast.LENGTH_LONG).show();
                    loadToast.error();
                }
            }, 1000);
        } else {
            getApp().getVaporizerCommunicator().setUpdateListener(this);
        }
        onUpdate(getApp().getVaporizerCommunicator().getData());
    }

    private App getApp() {
        return (App) getApplication();
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_settings:
                startActivity(new Intent(this, SettingsActivity.class));
                break;
            case R.id.action_help:
                new AlertDialog.Builder(this).setView(new HelpViewProvider().getView(this)).setTitle("Information").setPositiveButton("OK", null).show();
                break;

        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onUpdate(final VaporizerData data) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (introText.getVisibility() == VISIBLE) {
                    if (data.hasData()) {
                        introText.setVisibility(GONE);
                        loadToast.success();
                    } else {
                        introText.setText(Html.fromHtml(getString(R.string.intro_text)));
                        introText.setMovementMethod(new LinkMovementMethod());
                    }
                }

                vaporizerDataBinder.bind(data);
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return super.onCreateOptionsMenu(menu);
    }
}

File: mobile/src/main/java/org/ligi/vaporizercontrol/ui/DataDisplayActivity.kt
package org.ligi.vaporizercontrol.ui

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.text.Html
import android.text.method.LinkMovementMethod
import android.view.Menu
import android.view.MenuItem
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.WindowManager
import android.widget.Toast
import fr.nicolaspomepuy.discreetapprate.AppRate
import fr.nicolaspomepuy.discreetapprate.RetryPolicy
import kotlinx.android.synthetic.main.activity_main.*
import net.steamcrafted.loadtoast.LoadToast
import org.ligi.compat.HtmlCompat
import org.ligi.tracedroid.sending.TraceDroidEmailSender
import org.ligi.vaporizercontrol.R
import org.ligi.vaporizercontrol.VaporizerDataBinder
import org.ligi.vaporizercontrol.model.VaporizerData
import org.ligi.vaporizercontrol.wiring.App

class DataDisplayActivity : AppCompatActivity(), VaporizerData.VaporizerUpdateListener {

    private val loadToast by lazy { LoadToast(this) }
    private val vaporizerDataBinder by lazy { VaporizerDataBinder(this, app.settings) }
    private val app by lazy { application as App }

    private var lastNightMode: Int? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        findViewById(R.id.led)!!.setOnClickListener {
            val isUnknownOrNotBright = app.vaporizerCommunicator.data.ledPercentage == null || app.vaporizerCommunicator.data.ledPercentage === 0
            app.vaporizerCommunicator.setLEDBrightness(if (isUnknownOrNotBright) 100 else 0)
        }

        findViewById(R.id.led)!!.setOnLongClickListener {
            ChangeDialogs.showLEDPercentageDialog(this, app.vaporizerCommunicator)
            true
        }

        fab.setOnClickListener { fam.toggle() }

        findViewById(R.id.tempBoost)!!.setOnClickListener {
            ChangeDialogs.setBooster(this, app.vaporizerCommunicator)
        }

        fab_action_edit_boost.setOnClickListener {
            ChangeDialogs.setBooster(this, app.vaporizerCommunicator)
            fam.collapse()
        }

        fab_action_edit_settemp.setOnClickListener {
            ChangeDialogs.showTemperatureDialog(this, app.vaporizerCommunicator)
            fam.collapse()
        }

        fab_action_edit_led.setOnClickListener {
            ChangeDialogs.showLEDPercentageDialog(this, app.vaporizerCommunicator)
            fam.collapse()
        }

        findViewById(R.id.temperature)!!.setOnClickListener {
            ChangeDialogs.showTemperatureDialog(this, app.vaporizerCommunicator)
        }

        findViewById(R.id.temperatureSetPoint)!!.setOnClickListener {
            ChangeDialogs.showTemperatureDialog(this, app.vaporizerCommunicator)
        }

        intro_text.text = HtmlCompat.fromHtml(getString(R.string.intro_text))
        intro_text.movementMethod = LinkMovementMethod()

        AppRate.with(this).retryPolicy(RetryPolicy.EXPONENTIAL).initialLaunchCount(5).checkAndShow()

        if (!TraceDroidEmailSender.sendStackTraces("ligi@ligi.de", this) && !app.vaporizerCommunicator.data.hasData()) {
            loadToast.setText("searching crafty")
            loadToast.show()
        }
    }


    override fun onResume() {
        super.onResume()
        val bluetooth = app.vaporizerCommunicator.bluetooth
        if (bluetooth == null) {
            Handler().postDelayed({
                Toast.makeText(this@DataDisplayActivity, "can not scan - no BT available", Toast.LENGTH_LONG).show()
                loadToast.error()
            }, 1000)
        } else {
            if (!bluetooth.isEnabled) {
                bluetooth.enable()
            }
            app.vaporizerCommunicator.setUpdateListener(this)
        }
        onUpdate(app.vaporizerCommunicator.data)

        if (lastNightMode !=null && lastNightMode !=app.settings.nightMode) {
            recreate()
        }

        lastNightMode = app.settings.nightMode
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> startActivity(Intent(this, SettingsActivity::class.java))
            R.id.action_help -> HelpDialogFragment().show(supportFragmentManager, "help")
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onUpdate(data: VaporizerData) {
        runOnUiThread {
            if (data.hasData()) {
                if (intro_text.visibility == VISIBLE) {
                    dataContainer.visibility = VISIBLE
                    intro_text.visibility = GONE
                    loadToast.success()
                }

                vaporizerDataBinder.bind(data)
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onAttachedToWindow() {
        window.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
    }

}
